src/utils/config.py

```diff
@@ -0,0 +1,246 @@
+                isinstance(config1[key], dict)
+            ):
+                for subkey in config1[key]:
+                    if subkey not in config2[key].keys():
+                        config2[key][subkey] = config1[key][subkey]
+                    else:
+                        raise ValueError(f"Overlapping keys in configs: {key}:{subkey}")
+            else:
+                raise ValueError(f"Overlapping keys in configs: {key}")
+
+        else:
+            config2[key] = config1[key]
+    return deepcopy(config2)
+
+
+def _check_has_schema(value: dict) -> bool:
+    """Check if an item has a schema.
+
+    Args:
+        value (dict): The item.
+
+    Returns:
+        bool: Whether or not the item has a schema.
+    """
+    expected_keys = [
+        "top_level", 
+        "bottom_level", 
+        "dtype", 
+        "accept_nonetype",
+    ]
+    if sorted(expected_keys) == sorted(list(value.keys())[:4]):
+        return True
+    return False
+
+
+def _validate_path(path: str, param_nm: str, config: dict):
+    """Validate a passed path (str format) is valid.
+
+    Args:
+        path (str): The path to validate.
+        param_nm (str): The param name (for error raises).
+        item_conf (dict): The config file (for file ext).
+    """
+    _validate_str_is_path(path, param_nm)
+    try:
+        file_ext = config["filetype"]
+    except KeyError:
+        file_ext = None
+    if file_ext:
+        _validate_file_extension(path, file_ext)
+
+
+def _validate_numeric(value: Union[float, int], param_nm: str, config: dict):
+    """Valdiate a numerical value fits between a range.
+
+    Args:
+        value (Union[float, int]): The numerical value.
+        param_nm (str): The parameter name (used for error raises).
+        config (dict): The config.
+
+    Raises:
+        ValueError: Raised if a number is lesser than the minimum.
+        ValueError: Raised if a number is greater than the maximum.
+    """
+    try:
+        max = config["max"]
+    except KeyError:
+        max = None
+    try:
+        min = config["min"]
+    except KeyError:
+        min = None
+    if max:
+        if value > max:
+            raise ValueError(
+                f"Config value for {param_nm} ({value}) greater than max ({max})."
+            )
+    if min:
+        if value < min:
+            raise ValueError(
+                f"Config value for {param_nm} ({value}) less than min ({min})."
+            )
+
+
+def _nulltype_conversion(value: str) -> Union[str, None]:
+    """Account for user disparities in the form of NoneType they pass.
+
+    Args:
+        value (str): The value passed by a user.
+
+    Returns:
+        Union[str, None]: The altered value.
+    """
+    _type_defence(value, "value", str)
+    if value.lower() in ["", "null", "none"]:
+        return None
+    return value
+
+
+def _check_items(item: dict, config_item: dict, item_name: str) -> None:
+    """Check items of a config to validate them.
+
```

Comment:
I know this took some thinking through, so well done. However, it's not easy to follow, so could we have a description in the docstring to explain what's going on?

Reply:
added a small description